klib/gcp.c

```diff
@@ -698,6 +702,57 @@ static void gcp_metrics_add_memory(buffer body, boolean first, const char *type,
                   first ? "" : ",", gcp.zone, gcp.instance_id, type, state, value, interval);
 }
 
+static void gcp_metrics_add_disk(buffer body, boolean first, const char *type, const char *state,
+                                   u64 value, const char *interval, const char *device)
+{
+    bprintf(body, "%s{\"resource\":{\"type\":\"gce_instance\""
+                  ",\"labels\":{\"zone\":\"%b\",\"instance_id\":\"%b\"}}"
+                  ",\"metric\":{\"type\":\"agent.googleapis.com/disk/%s\""
+                  ",\"labels\":{\"device\":\"%s\",\"state\":\"%s\"}}"
+                  ",\"points\":[{\"value\":{\"doubleValue\":%ld},\"interval\":%s}]}",
+                  first ? "" : ",", gcp.zone, gcp.instance_id, type, device, state, value, interval);
+}
+
+closure_function(3, 4, void, gcp_metrics_disk_vh,
+                 buffer, b, const char *, interval, boolean, include_readonly,
+                 u8 *, uuid, const char *, label, filesystem, fs, inode, mount_point)
+{
+    if (filesystem_is_readonly(fs) && !bound(include_readonly))
+        return;
+    int block_size = fs_blocksize(fs);
+    u64 total_blocks = fs_totalblocks(fs);
+    u64 free_blocks = fs_freeblocks(fs);
+    u64 percent_used_free = free_blocks * 100 / total_blocks;
+    buffer b = bound(b);
+    const char *interval = bound(interval);
+    const char *device = 0;
+    if (label[0]) {
+        device = label;
+    } else if (uuid) {
+        buffer b_uuid = allocate_buffer(gcp.h, 2 * UUID_LEN + 4);
```

Comment:
Is this correct/recommended buffer size in this case
- `2(hex string) * 16(UUID_LEN) + 4(dashes) = 36 bytes`
- `9e7db206-f94d-f0ca-0c98-dd1c404d47a8`? 

What is the rationale/best practice behind allocation sizes, since one way or another it ends up calling `buffer_extend`

Reply:
Appreciate you taking the time to explain. Changes pushed. Thank you.